### PR TITLE
libc++3.7: fix to use with clang 3.9

### PR DESCRIPTION
**Pull request by veprbl**
###### Motivation for this change

Modern compiler will issue a following error whenever '#include <string>' is done:

```
/nix/store/yxpwamjdapjcp53mmsdh1j2c9bc26h4k-libc++-3.7.1/include/c++/v1/string:1938:44:
error: 'basic_string<_CharT, _Traits, _Allocator>' is missing exception specification 'noexcept(is_nothrow_copy_constructible<allocator_type>::value)'
basic_string<_CharT, _Traits, _Allocator>::basic_string(const allocator_type& __a)
                                           ^
/nix/store/yxpwamjdapjcp53mmsdh1j2c9bc26h4k-libc++-3.7.1/include/c++/v1/string:1326:40:
note: previous declaration is here
    _LIBCPP_INLINE_VISIBILITY explicit basic_string(const allocator_type& __a)
                                       ^
1 error generated.
```

This happens because modern clang is more strict about checking exception specification for forward declaration and definition.

This implements a part of upstream r242056. This affects newer version of CERN ROOT, which contains clang39-based interpreter cling.

###### Things done

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by jagajaga** [comment]
Don't forget to edit https://github.com/veprbl/nixpkgs/blob/2439f19df0c49e04fb3c9822d6c2009405c539bf/pkgs/misc/vim-plugins/vim2nix/additional-nix-code/youcompleteme

**Comment by veprbl** [comment]
@jagajaga gotcha!

**Comment by LnL7** [comment]
It's not really clear to me how the changes to `libc++3.7` relate to llvm/clang 3.9. (youcompleteme uses llvmPackages_39)

**Comment by veprbl** [comment]
nix builds compilers that don't remember about their "native" standard libraries. The path to the library is set by cc-wrapper instead. So putting llvm/clang 3.9 into ```buildInputs``` seems to be not enough to enable libc++3.9. And then there is a question whether we want to use libc++ 3.9 because YCM in principle depends on boost, which on nix darwin stdenv is built against libc++ 3.7, so there could be C++ ABI incompatibilities. Other significant dependency is libclang from clang 3.9, which seem to have only C ABI, so that is safe.

**Comment by veprbl** [comment]
This made me realize that we don't need clang 3.9 to build YCM. If should just compile with stdenv's gcc/clang.

**Comment by LnL7** [comment]
Oh, I thought youcompleteme really needed clang-3.9.

File: pkgs/misc/vim-plugins/default.nix [hunk 1]
```diff
@@ -1309,27 +1309,29 @@ rec {
...
     ] ++ stdenv.lib.optional stdenv.isDarwin Cocoa;
 
     propagatedBuildInputs = stdenv.lib.optional (!stdenv.isDarwin) rustracerd;
 
     patches = [
-      ./patches/youcompleteme/1-top-cmake.patch
       ./patches/youcompleteme/2-ycm-cmake.patch
     ];
 
+    # YCM requires path to external libclang 3.9
```
Comment (veprbl) [inline pkgs/misc/vim-plugins/default.nix]: But this is true
https://github.com/Valloric/YouCompleteMe/blob/master/README.md
```
YCM is designed to work with libclang version 3.9 or higher.
```

File: pkgs/development/compilers/llvm/3.7/libc++/default.nix [hunk 2]
```diff
@@ -14,7 +14,10 @@ stdenv.mkDerivation rec {
     cmakeFlagsArray=($cmakeFlagsArray -DLIBCXX_CXX_ABI_INCLUDE_PATHS="$NIX_BUILD_TOP/libcxxabi-${version}.src/include")
   '';
 
-  patches = [ ./darwin.patch ];
+  patches = [
+    ./darwin.patch
+    ./r242056.patch
```
Comment (veprbl) [inline pkgs/development/compilers/llvm/3.7/libc++/default.nix]: This is already fixed in 3.8. Fetchpatch didn't work for me, it complained about some filename. Also I had to manually edit paths in the original diff to change them from ```libcxx/trunk/...``` to ```a/...``` and ```b/...``` for compatibility with ```patch -p1```.

**Comment by veprbl** [comment]
Since YCM fix doesn't depend on this anymore, I moved it to PR #21574

**Comment by LnL7** [comment]
I think that explains my confusion.

**Comment by LnL7** [comment]
I'll keep an eye on the hydra build, the diff with master is pretty small at the moment so I can just merge it back into master if everything looks good.

**Comment by veprbl** [comment]
Good. Thank you!

**Comment by CMCDragonkai** [comment]
@veprbl You mentioned cling and I noticed that you packaged up the root data analysis framework. Does the root data analysis framework have cling installed as part of it, and is there a way to separate it out if you only want to use cling and not the root?

**Comment by veprbl** [comment]
@CMCDragonkai 
They say you can: https://root.cern.ch/cling
I have not tried myself though.

**Pull request merged**